### ISSUE TITLE
Loading status not showing

### ISSUE THREAD

**Issue by bhackett1024** [RecordReplay/devtools#5851]
I wrote a backend PR to cleanup the loaded / loading regions which are sent to the client (https://github.com/RecordReplay/backend/pull/4967), but this information is not shown appropriately while waiting for indexing to complete.

1. The loading progress indicator on the play button never updates while loading is happening.
2. Unloaded regions are not shown until after indexing completes.  I think it would be clearer if the unloaded regions reflect the latest information from the backend.

There are also a couple weird UI problems I noticed:

1. Most of the timeline is very indistinct and hard to see.
2. There is a spinner on the right which never goes away and has no tooltip, I don't know what it's for.

This replay shows all of these problems:

https://app.replay.io/recording/weird-loading-behavior--e09f6427-7141-455b-ba15-b54e6a96ca1f?point=217427430961111295803169762487312225&time=78448&hasFrames=false

**Comment by jonbell-lot23** [comment]
UI issues addressed in PR:

1. Contrast increased
2. Also fixed a bug that wasn't contrast related -- the class just wasn't loading

UI issues not in PR: The spinner is part of the video region, so it's as designed for now. 

--

And then the first two items are separate from my work. Will assign to Jaril to tackle the other stuff.

**Comment by bhackett1024** [comment]
There is more to this than the timeline contrast issue.

**Comment by jaril** [comment]
> Most of the timeline is very indistinct and hard to see.

I think this has been cleaned up in [Jon's PR](https://github.com/RecordReplay/devtools/pull/5986).

> There is a spinner on the right which never goes away and has no tooltip, I don't know what it's for.

It's originally meant to be there to indicate when the video playback is stalling. It looks like it's been co-opted as a loading indicator for when the mouse targets are loading in https://github.com/RecordReplay/devtools/pull/4832. Double checking that with @jasonLaster that doesn't look like the right place for that to me.

> Unloaded regions are not shown until after indexing completes. I think it would be clearer if the unloaded regions reflect the latest information from the backend.

That was intentional at first when started displaying unloaded regions to keep the UI from being too noisy, but I see how it can be confusing. @jonbell-lot23 I opened a new draft here where I turn off that behavior, so the timeline always reflects the current loaded/unloaded regions. Play with it and let me know what you think? https://github.com/RecordReplay/devtools/pull/6029

**Comment by jonbell-lot23** [comment]
@jaril Yeah, I don't recall the rationale for doing it the way it looks in prod, but I much prefer the #6029 draft you sent. I support moving to that.

**Comment by jaril** [comment]
> @jaril Yeah, I don't recall the rationale for doing it the way it looks in prod, but I much prefer the #6029 draft you sent. I support moving to that.

Let me clean it up and let's get it in!

**Comment by jaril** [comment]
@bhackett1024 just merged https://github.com/RecordReplay/devtools/pull/6029

Filing a separate PR for that spinner. It should only be showing specifically when your video is stuttering.

Will circle back and see if we can close this issue after then.

**Comment by jaril** [comment]
Last thing to address is:

> The loading progress indicator on the play button never updates while loading is happening.

The loading indicator around the play button is for the indexing progress. I think you might be expecting that it's there to indicate how long it will take for your unloaded regions in the timeline to become loaded? 

I need to do some more digging here. It looks like like have multiple loading concepts and I'm still not sure which one's which. For example, we keep track of these [two loading events](https://github.com/RecordReplay/devtools/blob/master/src/ui/actions/app.ts#L189-L192) when (I'm assuming) we process regions, but we also have the concept of [loadedRegions](https://github.com/RecordReplay/devtools/blob/master/src/protocol/thread/thread.ts#L297). 

Will update this issue with what I find!

**Comment by bhackett1024** [comment]
> @bhackett1024 just merged #6029

I actually had something else in mind here.  Would it make more sense to show the loading regions rather than the loaded regions on the timeline?  These will be relatively static so won't add visual noise during the loading state.  In that case as soon as we know which regions the backend is intending to load (the "loading" regions) we can highlight those on the timeline and the user will know what will be loaded when the progress indicator in the lower left finally finishes.

**Comment by bhackett1024** [comment]
> Last thing to address is:
> 
> > The loading progress indicator on the play button never updates while loading is happening.
> 
> The loading indicator around the play button is for the indexing progress. I think you might be expecting that it's there to indicate how long it will take for your unloaded regions in the timeline to become loaded?
> 
> I need to do some more digging here. It looks like we have multiple loading concepts and I'm still not sure which one's which. For example, we keep track of these [two loading events](https://github.com/RecordReplay/devtools/blob/master/src/ui/actions/app.ts#L189-L192) when (I'm assuming) we process regions, but we also have the concept of [loadedRegions](https://github.com/RecordReplay/devtools/blob/master/src/protocol/thread/thread.ts#L297).
> 
> Will update this issue with what I find!

Ah, that makes sense.  If the backend is returning inappropriate data for the unprocessed regions at the executionIndexed level then this is a backend problem.  In that case the progress indicator should show the fraction of the most recent "loading" regions which are also processed.  Alternatively we could show the fraction of the "loading" regions that are also "loaded", which is a similar concept and is simpler but might might lag behind the indexed regions by a few seconds.

**Comment by jaril** [comment]
> > @bhackett1024 just merged #6029
> 
> I actually had something else in mind here. Would it make more sense to show the loading regions rather than the loaded regions on the timeline? These will be relatively static so won't add visual noise during the loading state. In that case as soon as we know which regions the backend is intending to load (the "loading" regions) we can highlight those on the timeline and the user will know what will be loaded when the progress indicator in the lower left finally finishes.

I initially thought that'd be weird, but that actually makes a lot of sense. The "loaded" regions in the timeline is not really actionable info for the user — it's rare that they know exactly what they're looking for, and for them to expect that it will be withheld until the region it is in is loaded. 

Instead, the more relevant information is the loading region, since that's the part where we're saying "hey we'll have everything here ready for your at t time". We don't need the granularity of which parts have been loaded. The user just needs to know whether, within the loading region (regions we intend to load), they can start debugging things. And a (working and accurate) 0-100% progress indicator is a more useful signal there.

Questions:
- @jonbell-lot23 Thoughts? Happy to discuss sync later as well.
- @bhackett1024 This means we'll lose the visual of the loaded regions in the timeline. I think the user's largely unaffected. Does that change get in the way of your debugging-replay workflow, though?

> In that case the progress indicator should show the fraction of the most recent "loading" regions which are also processed.

The play button's progress indicator is currently used to signal the `executionIndexed` progress. If we were to replace that and show the loaded/loading (loaded over loading %) progress instead, is that good enough? I suppose my question is, is it still important that the frontend know about indexing? Or is it safe to say that everything you can do once indexing is finished, you'll also be able to do once loaded/loading is at 100%?

**Comment by bhackett1024** [comment]
>     * @bhackett1024 This means we'll lose the visual of the loaded regions in the timeline. I think the user's largely unaffected. Does that change get in the way of your debugging-replay workflow, though?

No, I think it's better to show the loading regions instead of the loaded regions in the timeline.

> The play button's progress indicator is currently used to signal the `executionIndexed` progress. If we were to replace that and show the loaded/loading (loaded over loading %) progress instead, is that good enough? I suppose my question is, is it still important that the frontend know about indexing? Or is it safe to say that everything you can do once indexing is finished, you'll also be able to do once loaded/loading is at 100%?

I think it would be fine to show the loaded / loading percentage.  When loading finishes execution indexing will not have finished yet, but it won't be far behind, typically a few seconds and should only extremely rarely be more than 10 seconds.

**Comment by bhackett1024** [comment]
> I think at the very least we should make sure that the indexing progress is correct. There are times when I would expect indexing to go from 100, and reset to not 100, then race back up to 100 again (e.g. editing the focus such that you load a previously unloaded region), but the UI doesn't appear to reflect that.

This would be good, and using the loaded / loading metric for this shouldn't be hard because any time the loading state changes in the backend (including after editing the focused range), these events will be emitted and the indicator can be updated.

> The thing I'm not sure about is—is it problematic to send an analysis request to the backend while indexing isn't complete? Or is it queued up properly so that you end up getting the right analysis results at the time it's fully indexed?

This is not problematic, the backend will wait until indexing finishes.

**Comment by jaril** [comment]
Discussion around onUnprocessedRegions:
https://github.com/RecordReplay/backend/issues/5123
